### PR TITLE
Cache URI Resolver

### PR DESCRIPTION
**Pull request by krisbitney**
This PR adds an implementations of a Result Resolution Cache URI Resolver, a PackageToWrapper Resolver, and a Wrapper Cache Resolver

### CODE REVIEW

**Comment by krisbitney** [comment]
> Looks good overall.
> 
> BTW, why not implement the `ResolutionResultCacheResolver` instead? We moved away from the `PackageToWrapperCacheResolver` and `WrapperCacheResolver` in the JS Client because of different issues with them.

Yes, sure, I will do this. Sounds good.

File: packages/resolvers/src/wrapper_cache_resolver.rs [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    },
+    error::Error,
+    wrapper::Wrapper,
+};
+use crate::cache::basic_wrapper_cache::BasicWrapperCache;
+use crate::cache::wrapper_cache::WrapperCache;
+use crate::uri_resolver_aggregator::UriResolverAggregator;
+
+/// A URI resolver that uses a cache to store and retrieve the results of resolved URIs.
+pub struct WrapperCacheResolver {
```
[the remark this answers is not in the document]
Comment (krisbitney) [reply packages/resolvers/src/wrapper_cache_resolver.rs:18]: I split it into two resolvers

File: packages/resolvers/src/resolution_result_cache_resolver.rs [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+            );
+            return result;
+        }
+
+        let sub_context = resolution_context.lock().unwrap().create_sub_history_context();
+        let sub_context = Arc::new(Mutex::new(sub_context));
+        let result = self.resolver.try_resolve_uri(uri, invoker.clone(), sub_context.clone());
+
+        if result.is_ok() {
+            self.cache_resolution_path(sub_context.clone(), Arc::from(result.clone()));
```
Comment (nerfZael) [inline packages/resolvers/src/resolution_result_cache_resolver.rs:82]: We should only cache the current uri here, not the whole path.
That way we keep the resolution path the same for cached and non cached wraps.
(This fixes issues with env which is the main reason we have this resolver)

**Review by nerfZael — APPROVED** [review]
Looks awesome, thanks for pushing this forward!

**Pull request merged**